### ISSUE TITLE
Permission Denied - Using Code Block Element when Wordfence is active.

### ISSUE THREAD

**Issue by danramzdev** [soflyy/breakdance-bugs#608]
Apparently there is a permissions error when adding code inside the php with the code block element when the Wordfence plugin is activated with its default configuration.

The message appears:

Permission Denied
Not enough permissions to perform the action.

**Comment by danramzdev** [comment]
I reinstalled Wordfence deleting the configuration I had, apparently that solved the problem, now having Wordfence active Breakdance works without sending the error.

**Comment by obedparla** [comment]
If Wordfence is setup to not allow `html` or `php` in ajax requests, there's nothing we can do to prevent it, since we need to send the data from the builder to the site. 

If you managed to work around it, Wordfence probably has a setting to turn that off

**Comment by JeusuDigital** [comment]
@danramzdev I just ran into the same issue, what you can do is go to the live traffic menu item in Wordfence and you will see the block in there, click the view icon and it has an option to Add Param to firewall allowlist, after doing this they rendered correctly for me.

**Comment by JeusuDigital** [comment]
Quick update, the above solution only seems to work per URL, so when trying to change URL's in the builder it gets blocked again, looking for a more permanent solution, only one currently working for me is adding your IP to Allowlisted IP addresses that bypass all rules in advanced firewall options, not ideal though as IP's generally change.

**Comment by JeusuDigital** [comment]
I have sent a request to the support team regarding the issue as it still exists and adding in IP every time it changes or expecting all site editors to do this is not ideal. I have found a temporary solution which is to append this to the bottom of the file in:

/wp-content/wflogs/rules.php

`$this->whitelistedParams['request.body[properties]'][] = '/.*/';`

Though this will likely be overwritten when it updates.

**Comment by JeusuDigital** [comment]
On further inspection into this issue I have found out that it is not specific to the code block but to SVG code.

So if I insert a code block with a standard console log in it or similar then no blocks in Wordfence, now if I insert a code block and then paste a simple SVG code in it like this:

`<svg height="100" width="100">
  <circle cx="50" cy="50" r="40" stroke="black" stroke-width="3" fill="red" />
  Sorry, your browser does not support inline SVG.  
</svg> `

It will be blocked immediately on render.

The issue is also not specific to the code block, it also happens on other blocks that have the ability to insert custom icons such as the social block, standard icon block, woocommerce mini cart etc.

**Comment by louisreingold** [comment]
When are the WordFence users going to tell the WordFence team to stop blocking our requests? 

Many people advise putting WordFence into learning mode.

**Comment by JeusuDigital** [comment]
@louisreingold I have notified them via support forum:
[https://wordpress.org/support/topic/wordfence-blocking-breakdance-page-builder/](https://wordpress.org/support/topic/wordfence-blocking-breakdance-page-builder/)

Hopefully they will look into it.

Learning mode will not work from my testing as when it puts the entries into allow it does so based on a URL, so you would need every single URL on your site loaded in there or you would still likely have blocks.

What it probably needs is for them to add exceptions based on the Ajax requests sent by the builder, something unique so it could bypass the standard rules that blocks things like SVG's being sent.

Or if they allowed wildcard in the URL that would work also like in the example above that I manually added to the rules file.

I'll do some more testing when I get the chance and update the thread if I get any further replies form them.

**Comment by louisreingold** [comment]
@JeusuDigital we added `_breakdance_doing_ajax=yes` to the URL of all Breakdance AJAX requests made from the builder.

Does WordFence provide some kind of parameter-based whitelist? If so, can you add the `_breakdance_doing_ajax` param to this whitelist so it lets in all requests that have `_breakdance_doing_ajax` with a value of `yes`?

**Comment by JeusuDigital** [comment]
@louisreingold Yes it does, the problem is in the plugin it does not allow wildcards on URL's only fixed URL's so we would need to add an exception for every page on the site as far as I can see. 

In their rules file (/wp-content/wflogs/rules.php) they have got wildcard rules based on any URL matching wildcards but the rules file updates and overwrites so not ideal adding it in there manually. 

What I think it needs is for them to add the exception / rule in the plugin file based on your new parameter being in the URL.

I have updated the support thread to let them know about the new parameter and asked if they can speak to you regarding whitelisting the Breakdance plugin.

**Comment by JeusuDigital** [comment]
Apparently they have finally added compatibility:
https://wordpress.org/support/topic/wordfence-still-blocking-breakdance-page-builder/

I can see the rule in there:

**/wp-content/wflogs/rules.php**

```
$this->whitelistedParams['request.body[tree]'][] = array(
'url' => '/\\/wp-admin\\/admin-ajax\\.php$/i',
'rules' => array (
  0 => '9',
),
'conditional' => new wfWAFRuleComparisonGroup(new wfWAFRuleComparisonGroup(new wfWAFRuleComparison($this, 'equals', 'breakdance_save', array(wfWAFRuleComparisonSubject::create($this, array('request.body', 'action'), array (
)))), new wfWAFRuleLogicalOperator('AND'), new wfWAFRuleComparisonGroup(new wfWAFRuleComparison($this, 'currentUserIs', 'administrator', array(wfWAFRuleComparisonSubject::create($this, 'server.empty', array (
)))), new wfWAFRuleLogicalOperator('OR'), new wfWAFRuleComparison($this, 'currentUserIs', 'editor', array(wfWAFRuleComparisonSubject::create($this, 'server.empty', array (
)))))))
);
```

I just had a block though even with that in there, I will need to test on a few more sites.

**Comment by plittlefield** [comment]
Any word on this?
